helix-term/src/commands.rs

```diff
@@ -1145,9 +1155,16 @@ fn search_impl(doc: &mut Document, view: &mut View, contents: &str, regex: &Rege
 
     // use find_at to find the next match after the cursor, loop around the end
     // Careful, `Regex` uses `bytes` as offsets, not character indices!
-    let mat = regex
-        .find_at(contents, start)
-        .or_else(|| regex.find(contents));
+    let mat = match direction {
+        Direction::Forward => regex
+            .find_at(contents, start)
+            .or_else(|| regex.find(contents)),
+        Direction::Backward => regex
+            .find_iter(&contents[..start])
+            .fold((None, None), |(_x, y), new| (y, Some(new)))
+            .0
+            .or_else(|| regex.find_iter(contents).last()),
```

Comment:
You should be able to optimize here by slicing `&contents[end..]` to only search for occurrences after the selection.